### PR TITLE
404 - changes to temporal queries

### LINKED ISSUE #404: Make accounts table temporal
Currently there is no way to view account history as Lorre updates account data in place in the database. We must therefore make the accounts table temporal. 

We can accomplish this by adding an `asof` column to the `accounts` table. Here is a caricature:

```
account_id | balance | asof | block_level
tz1kjdjk41o421kj2 | 20203437843 | 4pm | 500k
tz1kjdjk41o421kj2 | 20203437850 | 2pm | 475k
```

A query to get the latest balance for a given account would be of this form:

```sql
select * from accounts where account_id = 'tz1kjdjk41o421kj2' where asof<= now() order by asof desc limit 1;
```

A query to get the balance as of a given time would be of this form:

```sql
select * from accounts where account_id = 'tz1kjdjk41o421kj2' where asof<= 3 pm order by asof desc limit 1;
```

Finally, account history can be obtained with a simple query:

```sql
select * from accounts where account_id = 'tz1kjdjk41o421kj2' 
```

### PR DESCRIPTION
**Pull request by piotrkosecki**
Resolves #404 

Made changes to the the query, updated doc.

## Required changes to the DB

`accounts_history` table needs to change the `asof` column type because metadata couldn't recognize correct one:
```
asof timestamp without time zone NOT NULL
```

### CODE REVIEW

File: src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala [hunk 1]
```diff
@@ -41,20 +41,40 @@ object DataTypes {
...
+    val newPredicates =
+      fetchedAttributes.map { maybeAttributes =>
+        query.predicates.map { predicate =>
+          maybeAttributes.flatMap { attributes =>
+            attributes.find(_.name == dropAggregationPrefixes(predicate.field)).map {
+              case attribute if attribute.dataType == DataType.DateTime =>
+                predicate.copy(set = predicate.set.map(x => new Timestamp(x.toString.toLong).toString))
+              case _ => predicate
+            }
+          }.toList
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala]: Sorry, I'm having quite a hard time figuring out what's happening here, due to the Future[Option[List]] nesting and the interaction with the query optional values.

Could add a type annotation for all the `val`s  help understanding how the code behaves?

File: src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala [hunk 2]
```diff
@@ -277,14 +301,58 @@ object DataTypes {
...
   }
 
+  /** Class representing snapshot */
+  case class Snapshot(field: String, value: Any) {
+
+    /** Transforms snapshot to predicate */
+    def toPredicate: Predicate =
+      Predicate(
+        field = field,
+        operation = OperationType.after,
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala]: why is that the correct operation type?

File: src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala [hunk 3]
```diff
@@ -277,14 +301,58 @@ object DataTypes {
...
+  /** Finds invalid snapshot fields*/
+  def findInvalidSnapshotFields(
+      query: Query,
+      entity: EntityPath,
+      metadataConfiguration: MetadataConfiguration
+  ): List[InvalidSnapshotField] =
+    query.snapshot match {
+      case Some(snapshot) =>
+        metadataConfiguration.entity(entity).flatMap { entityCfg =>
+          entityCfg.attributes.filter {
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala]: Maybe I'm wrong here: is it filtering by value type and then looking for the right key?

Wouldn't it be simpler then to do a single `.find` that verifies both predicates, on key and value?

File: src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala [hunk 4]
```diff
@@ -277,14 +301,58 @@ object DataTypes {
...
+      )
+  }
+
+  /** Finds invalid snapshot fields*/
+  def findInvalidSnapshotFields(
+      query: Query,
+      entity: EntityPath,
+      metadataConfiguration: MetadataConfiguration
+  ): List[InvalidSnapshotField] =
+    query.snapshot match {
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala]: If you like to consider an alternative, you could try importing 
```scala
cats.instances.option._
cats.instances.list._

``` 

to have an extension `query.snapshot.foldMap` operation that would only need filling the mapping part, as the `List.empty` will be given for granted as the empty value for the list monoid.

File: src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperationsTest.scala [hunk 5]
```diff
@@ -3038,6 +3079,118 @@ class TezosDatabaseOperationsTest
 
       }
 
+      "should get the balance of an account at a specific timestamp where there are multiple entiries for given account_id" in {
```
Comment (ivanopagano) [inline src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperationsTest.scala]: typo: "entiries"

File: src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperationsTest.scala [hunk 6]
```diff
@@ -3038,6 +3079,118 @@ class TezosDatabaseOperationsTest
...
+
+        val ahrs = List(
```
Comment (ivanopagano) [inline src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperationsTest.scala]: more meaningful name? 😀

File: src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperationsTest.scala [hunk 7]
```diff
@@ -3038,6 +3079,118 @@ class TezosDatabaseOperationsTest
...
+          ),
+          Map(
+            "account_id" -> Some("id2"),
+            "block_id" -> Some("blockid2"),
+            "asof" -> Some(new Timestamp(2)),
+            "r" -> Some(1)
+          )
+        )
+
+      }
```
Comment (ivanopagano) [inline src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperationsTest.scala]: Maybe you might wanna test filtering failures on invalid field choice for the snapshot query field?

File: src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperationsTest.scala [hunk 8]
```diff
@@ -3038,6 +3079,118 @@ class TezosDatabaseOperationsTest
...
+        } yield found
+
+        val result = dbHandler.run(populateAndTest.transactionally).futureValue
+
+        result shouldBe List(
+          Map(
+            "account_id" -> Some("id1"),
+            "block_id" -> Some("blockid2"),
+            "asof" -> Some(new Timestamp(2)),
+            "r" -> Some(1)
```
[the remark this answers is not in the document]
Comment (piotrkosecki) [reply src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperationsTest.scala]: it's a rank in temporal query

File: src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperationsTest.scala (hunk 7 above)
Comment (piotrkosecki) [reply src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperationsTest.scala]: added test for that

File: src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperationsTest.scala (hunk 6 above)
Comment (piotrkosecki) [reply src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperationsTest.scala]: account history rows - changed

File: src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperationsTest.scala (hunk 5 above)
Comment (piotrkosecki) [reply src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperationsTest.scala]: changed

File: src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala (hunk 3 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala]: sure, changed

File: src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala (hunk 2 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala]: `after` is the same as `gt`, if you noticed it has also `inverse` set to `true`.
We want to tae the latest snapshot field by rank, and combining this predicate wit ordering we get that.

File: src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala (hunk 1 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala]: sure, done

File: src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala (hunk 4 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala]: done

**Pull request merged**